runtime/include/QuantumDevice.hpp

```diff
@@ -55,310 +57,558 @@ struct QuantumDevice {
      * calls.
      * Devices implementing support for this feature do not need to use the provided
-     * PRNG instance as their sole source of random numbers, but it is expected that the
-     * the same instance state will predictable and reproducibly generate the same
-     * program results. It is also expected that the provided PRNG state is evolved
-     * sufficiently so that two device executions sharing the same instance do not produce
-     * identical results.
-     * The provided PRNG instance is not thread-locked, and devices wishing to share it
-     * across threads will need to provide their own thread-safety.
-     *
-     * @param gen The std::mt19937 PRNG object.
+     * PRNG instance as their sole source of randomness, but it is expected that the
+     * the same instance state will predictably and reproducibly generate the same
+     * program results (considering all random processes in the device).
+     * It is also expected that the provided PRNG state is evolved sufficiently so that two copies
+     * of a device provided with the same PRNG instance do not produce identical results when run
+     * in succession.
+     * Note that the provided PRNG instance is not thread-locked, and devices wishing to use it
+     * across internal threads will need to provide their own thread-safety.
+     *
+     * @param gen Pointer to a Catalyst-managed Mersenne Twister instance.
      */
     virtual void SetDevicePRNG([[maybe_unused]] std::mt19937 *gen){};
 
+    // ----------------------------------------
+    //  QUANTUM OPERATIONS
+    // ----------------------------------------
+
+    /**
+     * @brief Apply an arbitrary quantum gate to the device.
+     *
+     * This instruction is opaque to the chosen quantum operation, which is specified via a
+     * string identifier. Additionally, an array of floating point parameters can be supplied,
+     * as well certain quantum modifiers, like the Hermitian adjoint (inverse) and control qubits.
+     * If the supplied combination of parameters is invalid or unsupported, the device should
+     * raise a runtime error.
+     *
+     * @param name A string identifier for the operation to apply.
+     * @param params Float parameters for parametric gates (may be empty).
+     * @param wires Qubits to apply the operation to.
+     * @param inverse Apply the inverse (Hermitian adjoint) of the operation.
+     * @param controlled_wires Control qubits applied to the operation.
+     * @param controlled_values Control values associated to the control qubits (equal length).
+     */
+    virtual void NamedOperation(const std::string &name, const std::vector<double> &params,
+                                const std::vector<QubitIdType> &wires, bool inverse = false,
+                                const std::vector<QubitIdType> &controlled_wires = {},
+                                const std::vector<bool> &controlled_values = {}) = 0;
```

Comment:
We can revisit the use of both `wires` and `qubits` in the interface. Do we need to keep both? Could we drop `wires` in favour of `qubits`?

Reply:
So it's true that I updated all descriptions to qubits (because I think that's clearer and more universal), but I didn't want to go in change all the variable names everywhere. Ideally we would only use one kind of name but I think it's fine.